### PR TITLE
elgg_view_input()

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds new api to generate form fields with a standard markup

``` html
<div class="elgg-field elgg-field-required">
    <label for="f1" class="elgg-field-label">Text input (.elgg-input-text):</label>
    <input value="input text" type="text" required="required" name="f1" id="f1" class="elgg-input-text">
    <div class="elgg-field-help elgg-text-help">This is how help text looks</div>
</div>
```

### CODE REVIEW

**Comment by hypeJunction** [comment]
* [ ] Where in the docs should this be mentioned?
* [ ] Should we use `::after` pseudo selector, or generate a `<span title="Required" class="elgg-?">*</span>`

File: engine/lib/views.php [hunk 1]
```diff
@@ -1269,6 +1269,57 @@ function elgg_view_form($action, $form_vars = array(), $body_vars = array()) {
 }
 
 /**
+ * Renders a form field
+ *
+ * @tip Field vars contain both field and input params. 'label', 'help', and 'field_class' params will
```
Comment (mrclay) [inline engine/lib/views.php]: I think this tip should be the inline doc for the $vars param. We should also indicate "label" and "help" accept HTML (they will not be escaped on output).

Comment (hypeJunction) [inline engine/lib/views.php]: Sounds good.

File: views/default/elements/forms/label.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+/**
+ * Form input label view
+ *
+ * @uses $vars['label'] Label text
```
Comment (mrclay) [inline views/default/elements/forms/label.php]: `@uses $vars['label'] HTML content of label element`

File: views/default/elements/forms/help.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+/**
+ * Form input help text view
+ *
+ * @uses $vars['help'] Help text
```
Comment (mrclay) [inline views/default/elements/forms/help.php]: Similarly here. Users should know it's HTML.

File: views/default/elements/forms/label.php (hunk 2 above)
Comment (hypeJunction) [inline views/default/elements/forms/label.php]: Of the <label> tag would be clearer I think

Comment (hypeJunction) [inline views/default/elements/forms/label.php]: Another alternative to allow either string or an array with 'text' parameter and other attributes to pass to the label tag

**Comment by mrclay** [comment]
We should update the admin CSS, too. I'm fine leaving this out of the default theme.

**Comment by hypeJunction** [comment]
One thing to consider is that in a lot of places core uses `<br />` after closing `</label>` tag. I think we should make `.elgg-field-label { display: block; }`

File: views/default/elements/forms/label.php (hunk 2 above)
Comment (mrclay) [inline views/default/elements/forms/label.php]: Eh keep it simple.

**Comment by mrclay** [comment]
Is there a use case in mind for the `elements/forms/input` view?

**Comment by mrclay** [comment]
Inlining just these labels seems fine to me.

**Comment by hypeJunction** [comment]
Yes, I want to be able to use 'view_vars' hook on all input elements. Currently, I would have to register a hook for each input view.
In my validation plugin, I will substitute 'required' param with `data-parsley-required`, to avoid issues with checkboxes and radios.

**Comment by mrclay** [comment]
Makes sense.

File: views/default/elements/forms/label.php (hunk 2 above)
Comment (hypeJunction) [inline views/default/elements/forms/label.php]: :ok:

**Comment by mrclay** [comment]
`<span title="Required" class="elgg-required-indicator"><i class="ui-helper-hidden-accessible">required</i></span>`, then add asterisk via `::after`?

**Comment by beck24** [comment]
:+1: 

Standardized form rendering will be a huge gain

Docs: http://learn.elgg.org/en/2.0/guides/actions.html
And would probably be good to update tutorials with standard forms: http://learn.elgg.org/en/2.0/tutorials/index.html

**Comment by hypeJunction** [comment]
@mrclay I settled for rendering an asterisk without any hidden tags. I think it will make it easier for plugin authors to overwrite and change, if needed. Pseudo element will raise questions where the asterisk is coming from.

@beck24 Updated docs.

File: docs/tutorials/blog.rst [hunk 4]
```diff
@@ -67,28 +67,34 @@ which we will create in a moment. Here is the content of
...
+    ]);
+
+    echo elgg_view_input('tags', [
+        'name' => 'tags',
+        'label' => elgg_echo('tags'),
+        'help' => elgg_echo('tags:help'),
+    ]);
+
+    echo elgg_view_input('submit', array(
+	    'value' => elgg_echo('save'),
```
Comment (juho-jaakkola) [inline docs/tutorials/blog.rst]: Mixed hard tab and spaces.

**Comment by juho-jaakkola** [comment]
Looks good! I'd even say let's merge this early and add tweaks and fixes in other PRs if necessary. (Assuming no-one finds actual blocking issues.)

**Comment by beck24** [comment]
Fantastic, LGTM!
Though this does open up a big chore to redo all existing forms this way and actually make everything consistent :)

**Comment by hypeJunction** [comment]
@juho-jaakkola Tab fixed

I have removed the `extend` view. There are already 3 extension points, and `extend` view would have the same effect as extending `help` view.

**Comment by hypeJunction** [comment]
I am hyped about these changes. Makes working with forms so much easier.

Two more plugins to illustrate what comes out it:
https://github.com/hypeJunction/elgg_forms_validation
https://github.com/hypeJunction/elgg_forms_register

**Comment by mrclay** [comment]
:+1:

**Comment by ewinslow** [comment]
> I am hyped about these changes.

I see what you did there.

**Comment by hypeJunction** [comment]
@ewinslow :nail_care:

**Pull request merged**